### DISCUSSION TITLE
How to share object between routes

### DISCUSSION THREAD

**Discussion by N7ghtm4r3** [Tlaster/PreCompose#396] [Q&A]
Hi, I'm trying to share a custom object between two routes with the `stateHolder`, but is the clean way to do that because sometimes the navigation starts after that the `stateHolder`  has been set the custom object, but sometimes, instead, the navigation starts before the custom object has been set. The following snippet is how I have implemented that:
```kotlin
MainScope().launch {
    val currentEntry = navigator.currentEntry.first()
    val stateHolder = currentEntry?.stateHolder
    stateHolder?.set("key", customObject)
}
navigator.navigate("route")
```

### Scene

```kotlin
scene(
   route = "route"
) { backstackEntry ->
      val stateHolder = backstackEntry.stateHolder
      val customObject: Any = stateHolder[key]!! // here the application crashes because the object has not been set when navigated here
      ...
}
```

May I have to wait before navigate? But if navigate inside the `MainScope` the navigation never happens

**Comment by Tlaster** [comment]
I would recommend don't share object between routes, I'd like to recommend follow the Android Architecture Guideline: https://developer.android.com/topic/architecture  
In short: It's better to follow "single source of truth", you can have a repository that store the object, and call it from different routes.

**Reply by N7ghtm4r3** [reply]
Thank you for the tip!

**Comment by N7ghtm4r3** [comment]
But if I need to do that in this case how can I do that?

**Comment by N7ghtm4r3** [comment]
I have followed your recommendation to implement a clean navigation
